Python: Code-only schema

helderco opened the pull request:
Fixes #3296

## What

No more need for crafting a `schema.graphql` file nor manually generating before commit. Python extensions can declare GraphQL schemas purely in code.

## How

You need a `server: dagger.Server` instance in a `main.py` file at the root of the extension.

```diff
# main.py
-if __name__ == '__main__':
-    s = dagger.Server()
-    s.run()
+schema = strawberry.Schema(query=Query)
+server = dagger.Server(schema)
```

## Playground

I wanted to keep the strawberry code as untouched as possible so you can use examples from the [docs](https://strawberry.rocks/docs) or from the [playground](https://play.strawberry.rocks/) with just a small addition, just like any [integration](https://strawberry.rocks/docs/integrations/aiohttp). 

Example from playground, you just copy the code do your extension and make a small change for the integration:

```diff
+import dagger
import strawberry

@strawberry.type
class User:
    name: str
    age: int

-@strawberry.type
+@strawberry.type(extend=True)
class Query:
    @strawberry.field
    def user(self) -> User:
        return User(name="Patrick", age=100)

schema = strawberry.Schema(query=Query)
+server = dagger.Server(schema)
```

## Other changes (hidden)

I took the opportunity to refactor some things, mainly to cover more use cases which are battle tested by adopted libraries:
- Delegating resolver execution to schema
- Delegating serialization to popular library (customizable via hooks)
- Adopted popular library to reduce class boilerplate[^1]
- Created executable that can be used as runtime entrypoint[^2]
- Simple custom logger
- Constrained dependency versions[^3]
- Formatted touched files with Prettier[^4]

[^1]: *attrs* is the popular library that originated the built-in *dataclasses* module.
[^2]: We can easily add more subcommands or arguments later on.
[^3]: It protects somewhat from major changes but we should adopt Poetry since Hatch doesn’t have reproducible builds yet.
[^4]: Plan for adding mypy, black and isort to dev dependencies for the SDK library.

## Async

There’s also async support in another PR:
- #3301

Signed-off-by: Helder Correia

helderco commented:
I'll have to rename the `dagger` python entrypoint because it'll conflict with the `dagger` CLI when run locally.

sipsma commented on sdk/python/dagger/src/dagger/cli.py:
```diff
@@ -0,0 +1,33 @@
+    sys.path.insert(0, ".")
+
+    try:
+        from main import server
```
huge nit: would it be possible to instead always import the `Query` class and then do the two lines of boilerplate for `schema = ...` and `server = ...` on behalf of the user?

helderco replied:
I'd rather not.

1. If you look at the [integrations](https://strawberry.rocks/docs/integrations/aiohttp), you always wrap the `schema` instance in your own system so I specifically wanted to keep this consistency. We're just building a Dagger integration here. Look at the "Playground" section in the description, as I explain it there.
2. You need control of `schema` for more stuff. If you have types not referenced directly but still need them in the schema for example, you need to reference them in this `Schema` constructor. I.e., there's more options there that you may need.
3. As for `server`, the point is to be able to configure it in some way.
    - Right now you can set `Server(..., debug=True)` for example. More options may be available later.
    - You can also add hooks to the converter to be able to serialize some complex object that you may have added (`server.converter.register_unstructure_hook(...)`). You can even instantiate `dagger.Server` with your own `Converter` object if needed.[^1]

[^1]: If this converter stuff looks undocumented, it's not. We just need to direct to the actual [library](https://cattrs.readthedocs.io/en/latest/usage.html) that does this.

As for this import, in case you think this is hacky, in Python it's actually not. It's pretty standard. It's normal in Python to have some config (i.e., "main:server" from file, env or cli arg) to get something like this. There's primitives to get an object from a string but for now I decided the more simpler approach (static), but we may allow it to be configurable if we want to (i.e., tell us where your `dagger.Server` instance is located).

sipsma replied:
Oh the only reason I asked here was to see if you thought it was easy to trim a little bit of boilerplate from the code the user has to write. If that constrains us in a bunch of other ways it's not a big deal at all, makes sense that we should continue letting them customize those fields (and stick close to the strawberry conventions when possible).

sipsma reviewed (commented):
I think we should add some integration tests to this.

One option would be to use the same sort of approach I used for the Go code-first support here: https://github.com/dagger/dagger/blob/7de43844ba577f3e7fba6ddbb2f2ff7a233d4718/core/integration/project_test.go#L83

helderco commented:
> I think we should add some integration tests to this.

Yes, I've mentioned this in https://github.com/dagger/dagger/pull/3310#issuecomment-1275952007. I'm prioritizing DX right now but plan on testing/stability right after.

helderco commented:
> One option would be to use the same sort of approach I used for the Go code-first support here:

I just saw what you mean with the `func TestCodeToSchema()` but the code to schema conversion in Python is already tested by the strawberry library that implemented it.

samalba commented on sdk/python/dagger/pyproject.toml:
```diff
@@ -4,24 +4,29 @@ build-backend = "hatchling.build"
 classifiers = [
     "Programming Language :: Python :: 3",
-#    "License :: OSI Approved :: MIT License",
+    "License :: OSI Approved :: Apple Public Source License",
```
Should it be Apache2 here?

helderco replied:
Ooops... great catch! I read it as "**Apache** Public Source License" 😅 I even looked for "Apache 2" specifically in https://pypi.org/classifiers/, found these:

- `License :: OSI Approved :: Apache Software License`
- `License :: OSI Approved :: Apache Public Source License`

And tried to figure out which was it. Thanks!

samalba reviewed (approved):
It's big cleanup, very nice work! I feel lazy seeing some of my old FIXMEs 🥲

I really like the pattern of the dagger.Server taking the schema, it's cleaner. Not sure about importing the `server` in the wrapper instead of starting it, but no strong opinion.

sipsma commented:
> I just saw what you mean with the func TestCodeToSchema() but the code to schema conversion in Python is already tested by the strawberry library that implemented it.

I really just meant that we should add a basic test that verifies Python extensions can load. I saw the other follow up issue you made for this though so that sounds good 👍

sipsma reviewed (approved):
When I pull the branch and run `dagger dev -p examples/python/hello/cloak.yaml ` I get this error while trying to obtain the schema:
```
#11 /entrypoint -schema
#11 0.069 + cd /src/examples/python/hello
#11 0.069 + exec dagger-python -schema
#11 0.069 + socat TCP-LISTEN:8080 UNIX-CONNECT:/dagger.sock
#11 0.069 /entrypoint: exec: line 8: dagger-python: not found
#11 ERROR: process "/entrypoint -schema" did not complete successfully: exit code: 127
```

Is that a known issue? Just double-checking. It's not a blocker right now as I know we're still iterating on all this. Feel free to merge if it's known+tracked.

helderco commented:
> Is that a known issue?

Yes, the problem is that the extension's requirements is using the merged version (in `cloak` branch):

```txt
git+https://github.com/dagger/dagger.git@cloak#subdirectory=sdk/python/dagger&egg=dagger
```

So after merge it should work as is.

I've had to do a few tweaks to make it work for development, that's why I created #3327, not just for those tools but in general.

Your comment makes me realize that at least a basic setup should be prioritized so others can try it in a PR without changes. I've just not thought about it much because the main thing is that I can't require an editable SDK library inside the extension because it runs in a container and it doesn't share the same tree.

Do you have any thoughts on that?

For now then, I've just been developing on a copy/clone of `sdk/python/dagger` → `examples/python/hello/dagger`, with following `requirements.txt`: `-e /src/examples/python/hello/dagger`.

helderco commented:
> Not sure about importing the `server` in the wrapper instead of starting it, but no strong opinion.

\- _Originally posted in https://github.com/dagger/dagger/pull/3299#pullrequestreview-1139854865_

Yeah, I was initially motivated by reducing boilerplate for running as a script. You can see in the [first commit](https://github.com/dagger/dagger/pull/3299/commits/1a5491bfc624f28533c72292ae4ff2b55397811a) that I replaced the `if __name__ == '__main__':` by a `__main__.py` file. Even then I wanted to have a default so we don't even require that extensions set any of those if they want something different.

At the same time I felt that the `execute()` method was doing too much:
- Parsing CLI args because of the `-schema` option
- Setting up the logger
- Running an async event loop: https://github.com/dagger/dagger/pull/3301

So in the [second commit](https://github.com/dagger/dagger/pull/3299/commits/48189d00a810ec594e569ce7c163c4f4a80e39cd) I moved all of that into a `cli.py` module which feels much cleaner and easily allows later options to the entrypoint or subcommands (e.g., `dagger-python server`, `dagger-python client`, `dagger-python generate`...).

Now I'm quite fond of this wrapper, I think it's cleaner and it's consistent with how other integrations for strawberry are made 🙂
